enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,509 @@
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+Introduce a new topology field in the Ingress config CR
+(`config.openshift.io/v1/ingresses`) called `IngressPlacement`.
+
+### Implementation Details/Notes/Constraints
+
+This new field will have one of these values - `ControlPlane` or `Workers`. 
+
+In addition, continue to allow the `.spec.replicas` and `.spec.nodePlacement`
+parameters in `operator.openshift.io/v1/ingresscontrollers` CRs to be omitted,
+but change the defaulting behavior for these fields.
+
+The value of the `IngressPlacement` field will affect the defaulting behavior
+of the `IngressController`'s `.spec.replicas` and `.spec.nodePlacement`
+parameters.  In the absence of an `IngressController` resource created by the
+user/installer, or when the user/installer creates an `IngressController` with
+these two parameters omitted, the Cluster Ingress Operator will choose the
+default values for those parameters based on the value of `IngressPlacement`.
+
+If the value of `IngressPlacement` itself is omitted by the installer, it
+is defaulted to `Workers`. This would be useful to maintain the current behavior
+if the API change is merged before the installer change to set this field, or if
+we want this newer OCP version to be backwards-compatible with older installers.
+
+When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
+`.spec.replicas` and `.spec.nodePlacement` will be the same as it is today:
+`.spec.replicas` will be chosen according to the value of
+`InfrastructureTopology`, namely `1` when `SingleReplica` or `2` when
+`HighlyAvailable`. `.spec.nodePlacement` will always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/worker: ''
+```
+
+However, if the value of `IngressPlacement` is `ControlPlane`, the defaulting
+behavior will be different: `.spec.replicas` will be chosen instead according
+to the value of `ControlPlaneTopology`; again, `1` when `SingleReplica` or `2`
+when `HighlyAvailable`. `.spec.nodePlacement` will be always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/master: ''
+```
+
+(Note that the `kubernetes.io/os: linux` label is mentioned just because it's
+the current behavior, it has no importance in this enhancement)
+
+The installer will detect situations in which it's unlikely the user will want
+to set up a load-balancer. For now, those situations include installation of
+single control-plane node cluster deployments on "on-prem" platforms such as
+"none" or "vSphere" (although today single control-plane node clusters are only
```

Comment:
Can we explicitly state which platforms will have this behavior and which will not? Is it **all** on-prem platforms? Also, I think it is a bit confusing to classify "none" as "on-prem". The "none" platform is agnostic and not necessarily on-prem.

Reply:
Yeah so I guess simply listing all the platforms is the way to go here

Reply:
The platforms are -
load balancer automatically provisioned / has VIP -
- alibaba
- aws
- azure
- gcp
- ibmcloud
- baremetal

load balancer needed by not provisioned -
- libvirt
- none
- openstack
- ovirt
- vsphere

Is that correct? Did I miss any platform?

Reply:
If this proposal is not applicable to platforms that use VIPs, then the "none" platform is the only platform to which this proposal applies.

For libvirt, the cluster uses local DNS for load balancing.

For openstack, ovirt, and vsphere, the cluster uses VIPs.

Reply:
Got it, revised.

Reply:
> I'm personally only really familiar with the installer behavior with cloud (AWS, azure, gcp, where a load balancer is provisioned by default) and the none platform (no load balancer provisioned by the installer by default), so I'm not the best person to answer this. @Miciah said there's also no load-balancer on vsphere...

I don't remember where or when I said that, but it's a little more nuanced as I understand it: vSphere UPI requires the user to set up an external LB; vSphere IPI sets up keepalived-ipfailover with a vip for ingress (and a vip for the API).  

> Yeah so I guess simply listing all the platforms is the way to go here

That's what we do in the IngressController API godoc.

Reply:
>I don't remember where or when I said that, but it's a little more nuanced as I understand it: vSphere UPI requires the user to set up an external LB; vSphere IPI sets up keepalived-ipfailover with a vip for ingress (and a vip for the API).

@staebler so would that mean we're gonna need more than a simple list of platforms here? A list of (platform,installation method) perhaps? 

Or would it be better to revert to not using an explicit list and to restore the phrasing of "Situations in which the installer provisions keepalived/LB"?

Reply:
The installer does not know the installation method (eg. IPI or UPI) that the user is going to take. As such, the installer cannot use the installer method as a criteria for configuring the Ingress CR manifest.

Ultimately, the installer needs code that will set the ingressPlacement in the Ingress CR manifest. I would prefer that the explicit list of platforms be stipulated here in the enhancement rather than deferring to later when it comes time to write the code to try to figure it out.

Reply:
>The installer does not know the installation method (eg. IPI or UPI) that the user is going to take. As such, the installer cannot use the installer method as a criteria for configuring the Ingress CR manifest.

In the case of vSphere, could we maybe be looking at `platform.vsphere.apiVIP` / `platform.vsphere.ingressVIP` (see [here](https://docs.openshift.com/container-platform/4.9/installing/installing_vsphere/installing-vsphere-installer-provisioned-customizations.html#installation-installer-provisioned-vsphere-config-yaml_installing-vsphere-installer-provisioned-customizations)) to determine the value of `IngressPlacement`?

Reply:
@omertuc This thread isn't resolved. Is the question answered in the body of the enhancement? If not, do we need an answer?

Reply:
The enhancement currently only refers to the none-platform, if we're interested in doing this enhancement on other platforms, I'll need the above question answered. This not coming from any real world use requirements, so I guess answering that is not urgent and could potentially be done in a separate enhancement. I wouldn't want to delay this enhancement further over this